### ISSUE TITLE
Implement multi-patch support in patch-package

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#20770]
### Problem

[patch-package](https://github.com/ds300/patch-package) is a simple and useful package that we use to implement bug fixes and new features in 3rd-party libraries without creating a fork. However, it's limited in that each package can have only one monolithic patch file. This makes it difficult to keep track of separate changes in a 3rd-party package, which is one of the reasons we've opted for forks in the past. You can track pull requests in the fork in parallel with their upstream equivalent and use that to reconcile all the custom changes we've made to a 3rd-party dependency. This also helps indicate when our code has landed upstream and the fork/patch is no longer needed.

### Solution
Upgrade patch-package to support multiple patch files per dependency. To ground this with an example, let's say we wanted to patch an image bug in React Native, and also wanted to build a new feature in the TextInput component.

For each of these features, we could have a separate patch file for each patch, and those patches would be applied in sequence:

```
patches/react-native+0.72.0+ImageBug.patch
patches/react-native+0.72.0+TextInputFeature.patch
```

We could have a separate PR in E/App to introduce each of these patches, and those PRs would link to their upstream equivalent.

Furthermore, we should have error reporting on a per-patch basis, and a way to resolve conflicts in specific patches while upgrading the base version.

#### Additional context

- [Original slack thread](https://expensify.slack.com/archives/C03TQ48KC/p1686161944375539)
- [Upstream issue](https://github.com/ds300/patch-package/issues/43)
- _Accounting note:_ Time spent on this should be considered R&D from a CAP SW perspective.

**Comment by roryabraham** [comment]
@shannonmccallexfy I'm assigning you here so that you can help make sure this project gets counted as CAP SW. I guess the `Accounting` auto-assigner doesn't work in E/App

**Comment by ds300** [comment]
Hi! 👋🏼 I'm David Sheldrick, the author of `patch-package`.

I had a call with @roryabraham a couple of days ago to hash out a prospective scope for this feature. He asked me to write up the following proposal:

## Scope

The project can broadly be broken down into three core features:

1. Patch sequence application.

   This would give the ability to load and apply multiple ordered patches to a single package, and for `patch-package` to convey to the user which patches failed, if any, during application. This would work the same way as cherry-picking a series of commits in git. The patch files should support semantic descriptions in the filenames, and might look something like:

    ```
    patches/react-native+0.72.0+01.ImageBug.patch
    patches/react-native+0.72.0+02.TextInputFeature.patch
    ```
    
   `patch-package` will attempt to apply all patches, even if one at the start or in the middle of the sequence fails to apply. And it will not leave merge conflicts in files. This matches the current single-patch behavior.
   
2. Patch sequence generation.

   This would give the ability to generate a new patch file on top of existing patch files, and to amend the 'latest' patch file. This would be achieved with an extra cli option, e.g.:

   - `patch-package react-native --append 'FixPressableEventDispatch'` to generate a new patch file with any extra changes added on top of existing patches.
      
     It would create a patch file named something like 

         patches/react-native+0.72.0+03.FixPressableEventDispatch.patch
         
   - `patch-package react-native` to amend the latest patch file (this syntax mimics the current patch generation behavior)
   
   The existing workflow of simply editing the relevant files inside `node_modules` and running `patch-package` to generate the patch file, whether to amend or append, would remain unchanged.
   
3. Error recovery via `--rebase`
   
   Let's say you have a sequence of five patches for `react-native` which build on one another. If the second patch in the sequence fails to apply after upgrading `react-native`, but you don't want to simply delete it, you should be given the opportunity to resolve the merge conflicts manually. `patch-package` can enable this by allowing you to run a command like
   
       patch-package react-native --rebase
   
   Which will
   
   - Undo any patches that were previously successfully applied
   - Re-apply the patches in order until one of them fails
   - Insert merge conflict markers into any files where conflicts were detected, as git does during merge/rebase.
   - Print a list of the files with conflicts, and the name of the patch that caused them.
   - Allow you to run `patch-package react-native --rebase-continue` to update the failing patch with the resolved changes and continue the rebase.

All features will have appropriate unit and integration tests, in line with the existing patch-package codebase.

### Integration with existing patch-package features

The [`--create-issue` flag](https://github.com/ds300/patch-package#options) will not be supported. 

The [`.dev.patch` patch file suffix](https://github.com/ds300/patch-package#dev-only-patches) _will_ be supported. This allows patch files created for transitive dev dependencies to be safely ignored if the package is not present in `node_modules`.

## Budget allocation

The total budget for this project is $20,000, and I'd propose allocating it between two milestones as follows

1. Patch sequence application + generation: **$15,000**
   These two features are the bulk of the work, and it doesn't make sense to ship one without the other so I'd propose combining them.
2. Error recovery via `--rebase`: **$5,000**

This would include a two-month period of asynchronous support + bug fixing while you battle-test the features, starting on the day of final delivery of milestone 2. I can also provide up to 2 hours of synchronous zoom support in the EST time zone over the same two-month period if needed, scheduled ad-hoc and subject to availability.

## Timeline

I expect to work on this in 1–2-hour chunks over the period of a month or so. I'll be working on it in my spare time during the height of summer so I can't commit to a fixed timeline, but I believe it's a realistic estimate. I'm obviously happy to keep you updated on my progress on a regular schedule (e.g. on Mondays and Thursdays, or whatever works for you), and I'll do the work in the open on the patch-package repo so you can follow along.

## Payment method

If convenient you can pay me via one-time GitHub sponsors donations, which I already have set up on the patch-package repo. Otherwise Rory mentioned that you normally use Upwork, which I am happy to register for.

Thanks,
David

**Comment by ds300** [comment]
Contributor details
Your Expensify account email: d.j.sheldrick@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~015445808e5ddd1173

**Comment by roryabraham** [comment]
> Patch sequence generation.

@ds300 What if you want to edit a specific patch file rather than the latest one? Maybe there should be three options:

- `npx patch-package my-package` – creates a new patch file if one does not exist, or edits the latest patch if one does exist
- `npx patch-package my-package --append 'PatchName'` – creates a new patch file to be applied after any existing patch files
- `npx patch-package my-package --update 'PatchName'` – looks for any patch file matching `PatchName`. Fails if no such file exists, otherwise it overwrites that specific patch file.

The use-case for that might be that you introduced a patch associated with an upstream PR, but then you make changes to the upstream and want to apply the same changes to your patch. Rather than creating a second patch associated with the same upstream PR, you would update the existing one.

> Integration with existing patch-package features

👍🏼 to both of the items there.

**Comment by roryabraham** [comment]
Overall, super excited for this feature, and thrilled that you're ready to build it for us and the whole community. I've assigned this issue to you, which means that we agree to your proposed scope and budget and you can start work whenever you're ready. Thanks! 🙇🏼

**Comment by ds300** [comment]
First Monday check-in! 🕐 

I haven't broken ground yet, planning to get started tonight after dinner.

> What if you want to edit a specific patch file rather than the latest one?

I think this would have to be part of the rebase tool, in order to make sure that it works for patches that build on one another sequentially. 

Just like git's rebase it could target a particular patch file in the 'history' and roll back to that point.

    patch-package react-native --rebase 'PatchName'

When you're done `--rebase-continue` would update the patch file you specified and then re-apply any subsequent patches.

**Comment by ds300** [comment]
The works is happening in this PR https://github.com/ds300/patch-package/pull/474

**Comment by mallenexpensify** [comment]
Thanks for the help with this @ds300 , I'll be managing payment.  Our normal process is to issue payment 7 days after PRs hit production, to ensure their aren't any regressions.  I'm unsure if that makes sense here though. Once we get closer to completion we'll discuss payment options, it'd be great to minimize any fees you might have to pay.

**Comment by ds300** [comment]
Thursday check-in! 🕐 

I have finished an initial implementation of patch sequence application and have done some testing for it. It seems to work well but in order to make it usable for patch generation and rebase it will need to be refactored to decouple the patch application bits from the error handling and console logging bits.

I plan to get started on that on Saturday morning.

**Comment by ds300** [comment]
Monday check-in! 🕐 

I got an initial implementation of patch sequence generation working. I need to improve the error reporting a little and do some more integration testing for it. After that I'll do some QA on different OSes, node versions, and package managers. After which I'll merge this PR and ping you. Probably some time later this week. Then I'll get started on the rebase feature.

I published a canary release of this branch on npm. You can do `npm install patch-package@canary` to try out the `--append` flag.

**Comment by ds300** [comment]
No changes since Monday. I expect i'll be able to work on it next on Monday

**Comment by ds300** [comment]
Monday check-in! 🕐 

I did some manual testing on all three platforms, then added and tested some sanity checks. I published a new canary and will merge the PR soon to begin work on the --rebase feature.

**Comment by ds300** [comment]
Nothing to report today. I'm planning to get started on --rebase on Saturday morning.

**Comment by roryabraham** [comment]
Thanks for the consistent updates @ds300!

**Comment by ds300** [comment]
Monday update 🕐 !

*tldr; ran into some unexpected complexity with patch sequence application. It is resolved now.*

While sitting down on Saturday to review the current state of the work and plan the --rebase feature I noticed that the new sequenced patch application code was no longer idempotent, and indeed would fail on subsequent attempts if patches build on top of one another in overlapping regions of code.

The patch application idempotency has so far been achieved by the following algorithm:

    try to apply patch
      if successful, we're done
      if not successful:
        reverse the patch and see whether it applies (do a dry run)
          if it does the patch was already applied
          if it does not the patch is corrupt or outdated
          
With a sequence of patches, doing it in this fashion (i.e. trying to 'dryly' apply the sequence backwards) would introduce a good bit of extra complexity.

Also, while thinking about the --rebase feature I realised it would require storing some state on disk the same way git does. So I opted to store a 'state' file already, in the patched package directory, e.g. `node_modules/react-native/.patch-package.json` which records the hashes of any applied patches the last time patch-package ran. It now checks the state file and the current patch file hashes to see whether or not it needs to run again.

We still use the old algo above if there is only one patch to apply, since it is battle-tested and naturally more robust due to being stateless.

I just finished this off and am now planning to get started on --rebase in earnest tomorrow. I have the day off on Wednesday and I'm planning to get the feature alpha testable by the end of the week.

I also just published a new canary with the idempotency fix.

**Comment by roryabraham** [comment]
FWIW it looks like we are going to take a stab at using this feature tomorrow. So @fabioh8010 if you have any feedback on the feature feel free to leave it here

**Comment by fabioh8010** [comment]
Hi 👋 I was able to use the `append` feature in this canary version 🎉 . I have one question, in case I want to "edit" a specific patch or the last one, is it possible to do it? I wonder if it's the case for the `rebase` option which is being developed, but I would want to confirm. Thanks!

**Comment by ds300** [comment]
@fabioh8010 that's great news! 🎉 

> I have one question, in case I want to "edit" a specific patch or the last one, is it possible to do it?

It's currently possible to edit only the _last_ patch, by making changes inside node_modules and running `patch-package react-native`.

The `--rebase` feature will allow you to edit any of the previous patches, or insert a new patch into the sequence.

🕐 I have nearly finished implementing an initial version of `--rebase`. I expect to release it in a canary tonight or tomorrow night. I will post here when it's available.

**Comment by ds300** [comment]
I just published a new canary with an initial version of the `--rebase` feature.

It still needs some edge case handling and testing but I think the bones are good now. Here's how it works:

## Updating an earlier patch

Let's say you have three patch files

  - `patches/react-native+0.72.0+001+FixPressableDispatch.patch`
  - `patches/react-native+0.72.0+002+FixTextSelection.patch`
  - `patches/react-native+0.72.0+003+FixDatePicker.patch`
  
If you want to update the middle one, you can run

    patch-package react-native --rebase patches/react-native+0.72.0+002+FixTextSelection.patch

This will undo the `003+FixDatePicker` patch. From that point, you can make any changes you need to inside `node_modules/react-native` and when you are finished, you can run.

    patch-package react-native
    
This will update the `002+FixTextSelection` patch file with your changes, then re-apply the `003+FixDatePicker` on top.

## Inserting a new patch into the sequence

Again, let's say you have the same three patch files, and you want to add a new one after `001+FixPressableDispatch`.

  - `patches/react-native+0.72.0+001+FixPressableDispatch.patch`
  - New one will go here
  - `patches/react-native+0.72.0+002+FixTextSelection.patch`
  - `patches/react-native+0.72.0+003+FixDatePicker.patch`
  
You can run

    patch-package react-native --rebase patches/react-native+0.72.0+001+FixPressableDispatch.patch
    
This will undo the `003+FixDatePicker` and `002+FixTextSelection` patches.

You are then free to make any changes you like inside `node_modules/react-native`. When you are done and you want to create the new patch file you can run:

    patch-package react-native --append 'FixTextRendering'

This will 
- create a new file called `patches/react-native+0.72.0+002+FixTextRendering.patch`
- rename the `002+FixTextSelection` and `003+FixDatePicker` patches to `003+FixTextSelection` and `004+FixDatePicker` respectively
- re-apply the `003+FixTextSelection` and `004+FixDatePicker` patches.

If you want to insert a patch at the start of the sequence you should be able to do `patch-package react-native --rebase 0`, though I haven't tested this yet.

Let me know if you have any questions.

**Comment by fabioh8010** [comment]
Hi @ds300 Thanks for your explanation and work on this feature!

Today I came across another situation where I had to remove the last patch applied to the library. I have 3 patches applied to the `react-native` library:

1. react-native+0.72.1+001+initial.patch
1. react-native+0.72.1+002+NumberOfLines.patch
1. react-native+0.72.1+003+VerticalScrollBarPosition.patch

I tried removing the last file and running `npm install`, but I got this kind of error:

```
TypeError: Cannot read properties of undefined (reading 'patchFilename')
    at Object.applyPatchesForApp (/Users/fabiohenriques/Documents/Work/Callstack/Projects/Expensify/expensify-app/node_modules/patch-package/dist/applyPatches.js:75:102)
    at Object.<anonymous> (/Users/fabiohenriques/Documents/Work/Callstack/Projects/Expensify/expensify-app/node_modules/patch-package/dist/index.js:77:24)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/Users/fabiohenriques/Documents/Work/Callstack/Projects/Expensify/expensify-app/node_modules/patch-package/index.js:3:1)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
```

So, I ended up having to remove the object related to `react-native+0.72.1+003+VerticalScrollBarPosition.patch` in `node_modules/react-native/.patch-package.json` to get rid of this error.

```json
{
  "version": 0,
  "patches": [
    {
      "patchFilename": "react-native+0.72.1+001+initial.patch",
      "patchContentHash": "c3dfb880c7386f6dfdcedb1a53453a6a1035208055462649a0b85b36e101c9c5",
      "didApply": true
    },
    {
      "patchFilename": "react-native+0.72.1+002+NumberOfLines.patch",
      "patchContentHash": "52e9be0f3e318e39921ddb8e60ef742d37950e228cf8c51c221ca00b5ca9065d",
      "didApply": true
    },
    { // <- I removed this object
      "patchFilename": "react-native+0.72.1+003+VerticalScrollBarPosition.patch",
      "patchContentHash": "d69af356dee71cd82b863995ca79f43ae0b2461ce9b74eeb00e6ca8cce240cd6",
      "didApply": true
    }
  ]
}
```

**Comment by ds300** [comment]
@fabioh8010 thanks for the bug report.

When removing a patch file you should reinstall node_modules from scratch rather than merely running `npm install`. Looks like I need to add extra integrity checks to the state file handling logic to let people know this fact.

**Comment by ds300** [comment]
Apologies for missing the check-in yesterday.

No changes since Friday. I'm aiming to reach beta for this feature by the end of the week. All that remains is adding and testing some edge cases for the `--rebase` flow, and then writing some docs.

**Comment by roryabraham** [comment]
@ds300 that's great news!

**Comment by ds300** [comment]
Monday Update!

> I also need to add support for optionally inserting conflict markers when a patch fails so I'm planning to do that tonight or on Saturday.

It turns out this isn't easily doable. Git can do it because it has knowledge of the full history of the file in both branches being merged, while we only have one side of that history in our patch files. We could theoretically download previous versions of the package and figure out diffs from that but it'd be a lot more work than I bargained for.

As a simple alternative I can make it apply as many hunks as it can, while writing a log of the hunks that failed to apply (along with clickable links to the relevant files). This should make it a bit easier to reapply the failing hunks manually.

What do you think?

**Comment by roryabraham** [comment]
That sounds like a pretty good alternative to me 👍🏼

**Comment by ds300** [comment]
Thursday update! Hopefully for the last time! 🤞🏼 

I added a `--partial` option which does the thing I described above, along with tests and docs.

I just released a new major version with this feature, 8.0.0. 

What's the situation on your end regarding testing/adoption?

**Comment by ds300** [comment]
@roryabraham any update from your side?

**Comment by roryabraham** [comment]
We have started using the feature in production! @fabioh8010 I know you mentioned you have some minor feedback on the feature, can you leave that feedback here please?

That said, I think the remaining issues are minor and we can issue payment for this task. @ds300, thanks so much for working on this! 🚀

**Comment by roryabraham** [comment]
@mallenexpensify @ds300 can you work together to complete the payment for this issue please?

**Comment by fabioh8010** [comment]
Hi @ds300 ! Yes, yesterday I used the `8.0.0` release and noticed 2 minor things, I will create each one in separate messages.

**Comment by fabioh8010** [comment]
**Strange behaviors when we have multi-patch files version mismatch**

I've upgraded RN to `0.72.3` and when running `npm install` I got some repeated messages about the patch file version mismatch.

```
> new.expensify@1.3.50-3 postinstall
> scripts/postInstall.sh

patch-package 8.0.0
Applying patches...
@oguzhnatly/react-native-image-manipulator@1.0.5 ✔
@onfido/react-native-sdk@7.4.0 ✔
@react-navigation/stack@6.3.16 ✔
react-native@0.72.1 (1 initial) ✔
react-native@0.72.1 (2 NumberOfLines) ✔
react-native@0.72.1 (3 VerticalScrollBarPosition) ✔
react-native-fast-image@8.6.3 ✔
react-native-modal@13.0.1 ✔
react-native-vision-camera@2.15.4 ✔
react-native-web-lottie@1.4.4 ✔

Warning: patch-package detected a patch file version mismatch

  Don't worry! This is probably fine. The patch was still applied
  successfully. Here's the deets:

  Patch file created for

    react-native@0.72.1

  applied to

    react-native@0.72.3
  
  At path
  
    node_modules/react-native

  This warning is just to give you a heads-up. There is a small chance of
  breakage even though the patch was applied successfully. Make sure the package
  still behaves like you expect (you wrote tests, right?) and then run

    patch-package react-native

  to update the version in the patch file name and make this warning go away.


Warning: patch-package detected a patch file version mismatch

  Don't worry! This is probably fine. The patch was still applied
  successfully. Here's the deets:

  Patch file created for

    react-native@0.72.1

  applied to

    react-native@0.72.3
  
  At path
  
    node_modules/react-native

  This warning is just to give you a heads-up. There is a small chance of
  breakage even though the patch was applied successfully. Make sure the package
  still behaves like you expect (you wrote tests, right?) and then run

    patch-package react-native

  to update the version in the patch file name and make this warning go away.


Warning: patch-package detected a patch file version mismatch

  Don't worry! This is probably fine. The patch was still applied
  successfully. Here's the deets:

  Patch file created for

    react-native@0.72.1

  applied to

    react-native@0.72.3
  
  At path
  
    node_modules/react-native

  This warning is just to give you a heads-up. There is a small chance of
  breakage even though the patch was applied successfully. Make sure the package
  still behaves like you expect (you wrote tests, right?) and then run

    patch-package react-native

  to update the version in the patch file name and make this warning go away.

---
patch-package finished with 3 warning(s).

up to date, audited 47 packages in 494ms

4 packages are looking for funding
  run `npm fund` for details

1 moderate severity vulnerability

To address all issues, run:
  npm audit fix

Run `npm audit` for details.

added 3960 packages, and audited 3961 packages in 1m

398 packages are looking for funding
  run `npm fund` for details

43 vulnerabilities (15 moderate, 25 high, 3 critical)

To address issues that do not require attention, run:
  npm audit fix

To address all issues possible (including breaking changes), run:
  npm audit fix --force

Some issues need review, and may require choosing
a different dependency.

Run `npm audit` for details.
~/Documents/Work/Callstack/Projects/Expensify/expensify-app-rn-update (feature/rn-update-0.72.3) $
```

As you can see I got three equal warnings about the mismatch with the same instructions to fix it.

When running `npx patch-package react-native`, instead of renaming my patch files it created a new one based on the last patch file.

```
~/Documents/Work/Callstack/Projects/Expensify/expensify-app-rn-update (feature/rn-update-0.72.3) $ npx patch-package react-native
patch-package 8.0.0
• Creating temporary folder
• Installing react-native@0.72.3 with npm
• Diffing your files with clean files
✔ Created file patches/react-native+0.72.3+003+VerticalScrollBarPosition.patch

~/Documents/Work/Callstack/Projects/Expensify/expensify-app-rn-update (feature/rn-update-0.72.3) $
```

So I ended up having a corrupt multi-patch setup:

<img width="499" alt="Screenshot 2023-08-08 at 14 56 52" src="https://github.com/Expensify/App/assets/20051562/25626386-333d-466f-b05c-788a3cd1b922">


To fix it I had to delete `node_modules` and rename the patch files manually.

**Comment by mallenexpensify** [comment]
Sorry @roryabraham and @ds300 , I was outta commission for a couple weeks. 

@roryabraham can you please confirm the payment is $20k? I'm basing it on [this post](https://github.com/Expensify/App/issues/20770#issuecomment-1593560155)

Has payment method been discussed? @ds300 do you preferences? 

Bumping to daily to expedite til paid

**Comment by ds300** [comment]
@mallenexpensify thanks for picking this up! upwork.com seems like the best payment option for me.

@fabioh8010 that's really useful feedback thanks. I'll plan to fix those issues soon and will report back here.

**Comment by ds300** [comment]
@mallenexpensify @roryabraham can we get this moving again?

**Comment by roryabraham** [comment]
Hey hey sorry for the delay. Yes @mallenexpensify $20,000 was discussed and agreed upon by both parties. [internal slack context](https://expensify.slack.com/archives/C03TQ48KC/p1686677926840699?thread_ts=1686161944.375539&cid=C03TQ48KC)

**Comment by roryabraham** [comment]
@shannonmccallexfy heads-up that this should be coded as R&D

**Comment by mallenexpensify** [comment]
@ds300 , can you please accept the job and reply here once you have? 
https://www.upwork.com/jobs/~0126b304cc4c67a455

Can you also include a screenshot of your Upwork profile to ensure I'm paying the right person? 

Thanks

**Comment by ds300** [comment]
this is my upwork profile
<img width="1076" alt="image" src="https://github.com/Expensify/App/assets/1242537/f0ea5f89-e81a-4eae-a87f-937a68a771e1">

**Comment by mobileteamt** [comment]
**Proposal**

**Please re-state the problem that we are trying to solve in this issue.**


extending patch-package to support multiple patches for the same package. This will allow users to apply and manage multiple patches to a single package in a straightforward and organized manner. Key features of the proposed solution include:


**What is the root cause of that problem?**

the implementation of multi-patch support in patch-package is the increasing complexity of modern software projects and the need for more granular and organized management of patches

 

**What changes do you think we should make in order to solve the problem?**


To implement multi-patch support in patch-package, the following technical steps should be taken:


Directory Structure: Create a dedicated directory structure within the project to store multiple patches for each package. For example:

 

	patches/

  		package-name/

   			patch1.diff

    			patch2.diff	

 

Command-Line Interface (CLI): Extend the patch-package CLI to include commands for applying, listing, and removing multiple patches for a package.

 

Patch Ordering: Implement logic to apply patches in the specified order, ensuring that each patch is applied relative to the package's original state.

 

Version Compatibility: Develop a mechanism to ensure that patches are compatible with different package versions and prevent conflicts when applying patches to packages with varying codebases.

 

Documentation: Update the patch-package documentation to reflect the new multi-patch functionality, providing clear instructions on how to use it effectively.

 

Testing: Create comprehensive unit tests and integration tests to verify the functionality of multi-patch support and ensure backward compatibility with existing single-patch use cases.


**What alternative solutions did you explore? (Optional)**

None.

**Comment by mallenexpensify** [comment]
@ds300 the plan is:
- Pay $12,000 USD (max allowable) for a sponsorship via https://github.com/sponsors/ds300
- Pay $8,000 USD balance via GH sponsorship (assuming I'm able to after the initial payment)

That sound good? I should be able to handle this week, I'm waiting for my card limit to be raised

**Comment by ds300** [comment]
Works for me, thanks @mallenexpensify 🙏🏼

**Comment by mallenexpensify** [comment]
@ds300 , I've checked with out accounting team and we need to pay you through Expensify.   

Contributor: @ds300 is due $20,000 via Expensify NewDot.  

@JmillsExpensify and/or @anmurali can you please issue payment?  What are the steps @ds300 needs to do in order to receive payment, it looks like he's in the UK.

**Comment by JmillsExpensify** [comment]
@mallenexpensify Can you add him to the Contributor Payments workspace owned by contributors@expensify.com? Then he'll need to `Request money` in his `policyExpenseChat`.

**Comment by mallenexpensify** [comment]
Thanks @JmillsExpensify I've added @ds300 via their gmail address we've communicated through before. I send these steps, let me know if anything doesn't look correct

- Big Green Plus button
- Request Money
- Manual > Amount ($20,000)
- [contributors@expensify.com](mailto:contributors@expensify.com)
- Enter https://github.com/Expensify/App/issues/20770 into description

Anyone know what the best payment setup is for someone in the UK?  Wise? Can we do direct transfer/payment to a bank account there?

**Comment by roryabraham** [comment]
Posted question in slack [here](https://expensify.slack.com/archives/C03U7DCU4/p1695213505162749)

**Comment by anmurali** [comment]
1. Please make sure you have added them to the Global Reimbursements beta though and to the `Expensify Contributors (Collect)` workspace in OD as well as `Expensify Contributors` workspace in ND
2. Please make sure they have their default workspace set to the `Expensify Contributors (Collect)` policy on Old Dot before going to the next step
3. Instructions for a contributor plus to add their deposit bank account to get paid is [here](https://expensify.slack.com/archives/C02NK2DQWUX/p1687191320448749?thread_ts=1686871517.189609&cid=C02NK2DQWUX)
4. Instructions for a contributor plus to request a payment is [here](https://expensify.slack.com/archives/C02NK2DQWUX/p1687970684314839?thread_ts=1686871517.189609&cid=C02NK2DQWUX)



>Anyone know what the best payment setup is for someone in the UK? Wise? Can we do direct transfer/payment to a bank account there?
They can add their UK bank account or a Wise account, it's their call.

**Comment by mallenexpensify** [comment]
I've been in touch with @ds300 and provided details. Once I've confirmed @ds300 is setup I'll comment here for payment. 

 I've updated internal documentation in [this SO](https://stackoverflowteams.com/c/expensify/questions/17415)

**Comment by mallenexpensify** [comment]
Contributor: @ds300 is due $20,000 via NewDot.  

@ds300 please confirm you've followed the steps to request payment that I sent you via email.

**Comment by JmillsExpensify** [comment]
$20,000 payment approved via NewDot based on BZ summary.

**Comment by ds300** [comment]
> @ds300 please confirm you've followed the steps to request payment that I sent you via email.

I confirm

**Comment by mallenexpensify** [comment]
Thanks @ds300.  Can you comment once you've received the payment then I'll close this.

**Comment by ds300** [comment]
I just received a notification saying that an ACH direct deposit was made, and I should have the money by Oct 4th.

Will post an update here once it's gone through.

**Comment by ds300** [comment]
The payment went through! Thanks for figuring this out @mallenexpensify and @roryabraham. It's been a pleasure to work with you both.

**Comment by roryabraham** [comment]
Great news, thanks for building this awesome feature for us. I hope the open-source community finds it as valuable as we do!

**Comment by mallenexpensify** [comment]
Thanks @ds300 , glad everything worked out!
